### PR TITLE
[master < T0878-MG] Add a python manual test script that verifies that we meet certain isolation levels

### PR DESCRIPTION
**Pull request by spacejam**
This script checks whether certain anomalies are correctly detected by our MVCC code, forcing conflicting transactions to abort as expected. Our system currently satisfies the requirements for Consistent View and Snapshot Isolation. This script can be used in the future whenever MVCC-related logic is updated, and it will be particularly useful when we start writing a distributed transaction processing system.

Output:

```
λ python tests/manual/test_isolation_level.py
✓ g0 test passed
✓ g1a test passed
✓ g1b test passed
✓ g1c test passed
✓ otv test passed
✓ pmp test passed
✓ pmp_write test passed
✓ p4 test passed
✓ g_single test passed
✓ g_single_dependencies test passed
✓ g_single_write_1 test passed
✓ g_single_write_2 test passed (although the abort rate is pessimistically high)
expected transaction to fail with conflicting transactions but it incorrectly committed successfully
X g2_item test failed - database exhibits write skew - writes based on invalidated reads should have failed, causing repeatable read (PL-2.99) and serializability (PL-3) to fail to be achieved
expected transaction to fail with conflicting transactions but it incorrectly committed successfully
X g2 test failed - database exhibits write skew on predicate read - concurrent transactions that should have caused a predicate read to return data in one transaction actually returned nothing in both. Both transactions committed, but one of them should have failed due to having a predicate invalidated
X g2_two_edges test failed: database exhibits write skew on predicate read - a transaction's read set was invalidated in two concurrent transactions,and it should have failed to commit if we want to be serializable

results:
consistent view (PL-2+): True
snapshot isolation (PL-SI): True
repeatable read (PL-2.99): False
full serializability (PL-3): False

more information about these anomalies can be found at:
        https://github.com/ept/hermitage
        https://pmg.csail.mit.edu/papers/adya-phd.pdf
```

### CODE REVIEW

**Review by 42jeremy — APPROVED** [review]

File: tests/e2e/test_isolation_level.py [hunk 1]
```diff
@@ -0,0 +1,568 @@
...
+    cursor.execute("""
+            // %s
+         MERGE (kv:Kv{ key: $key })
+           SET kv.value = $value
+        RETURN kv;
+    """ % comment, {'key': key, 'value': value})
+    ret = cursor.fetchall()
+    if len(ret) != 1:
+        print("expected ret to be len 1, but it's", ret)
+        assert(len(ret) == 1)
```
Comment (jbajic) [inline tests/e2e/test_isolation_level.py:55]:
```suggestion
        assert len(ret) == 1
```

File: tests/e2e/test_isolation_level.py [hunk 2]
```diff
@@ -0,0 +1,568 @@
...
+def update(cursor, key, value, comment = ""):
+    cursor.execute("""
+            // %s
+         MERGE (kv:Kv{ key: $key })
+           SET kv.value = $value
+        RETURN kv;
+    """ % comment, {'key': key, 'value': value})
+    ret = cursor.fetchall()
+    if len(ret) != 1:
+        print("expected ret to be len 1, but it's", ret)
```
Comment (jbajic) [inline tests/e2e/test_isolation_level.py:54]: Favor f-strings formatting, it makes reading code more intuitive and it's a new optimized and more favoured  way to format a strings
```suggestion
        print(f"expected ret to be len 1, but it's {ret}")
```

File: tests/e2e/test_isolation_level.py [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+def assert_commit_fails(conn, failure = "conflicting transactions"):
+    try:
+        conn.commit()
+        # should always abort
+        print("expected transaction to fail with", failure, "but it incorrectly committed successfully")
+        assert(False)
+    except mgclient.DatabaseError as e:
+        if not failure in str(e):
+            print("expected exception containing text", failure, "but instead it was", str(e))
+            assert(False)
```
Comment (jbajic) [inline tests/e2e/test_isolation_level.py:83]:
```suggestion
        assert failure in str(e), "expected exception containing text {failure} but instead it was {e}"
```

File: tests/e2e/test_isolation_level.py [hunk 4]
```diff
@@ -0,0 +1,568 @@
...
+    if actual != expected:
+        print("expected key", key, "to have value", expected, "but instead it had value", actual)
+        assert(False)
+
+def select_all(cursor, expected):
+    cursor.execute("MATCH (kv: Kv {}) RETURN kv")
+    actual = [ r[0].properties['value'] for r in cursor.fetchall() ]
+    if actual != expected:
+        print("expected values", expected, "but instead it had values", actual)
+        assert(False)
```
Comment (jbajic) [inline tests/e2e/test_isolation_level.py:105]:
```suggestion
    assert actual == expected, "expected values {expected}, but instead it had values {actual}"
```

File: tests/e2e/test_isolation_level.py [hunk 5]
```diff
@@ -0,0 +1,568 @@
...
+    setup(c1)
+    g2_two_edges = g2_two_edges(c1, c2, c3)
+
+    g1 = g1a and g1b and g1c
+    repeatable_read = g1 and g2_item
+    snapshot_isolation = g0 and g1 and otv and pmp and \
+        pmp_write and p4 and g_single and g_single_dependencies \
+        and g_single_write_1 and g_single_write_2
+    full_serializability =  snapshot_isolation and g2_item \
+        and g2 and g2_two_edges
```
Comment (jbajic) [inline tests/e2e/test_isolation_level.py:553]:
```suggestion
    g1 = all([g1a, g1b, g1c])
    repeatable_read = all([g1 , g2_item])
    snapshot_isolation = all([g0,  g1, otv,  pmp,
        pmp_write, p4, g_single, g_single_dependencies,
        g_single_write_1, g_single_write_2])
    full_serializability =  all([snapshot_isolation, g2_item,
        g2, g2_two_edges])
```

**Comment by vpavicic** [comment]
@gitbuda - do i need to add something to the changelog? What?

**Comment by gitbuda** [comment]
Nothing for changelog here 😄

**Pull request merged**